142_张嘉_第一课作业

mind-gravity opened the pull request:
我是插班生，今天下午才入学，所以这次作业晚了点交，请谅解。

mind-gravity committed:
第一课作业

1、复盘课件内容
2、传参实现可变地址可变薪水数

mind-gravity committed:
Merge pull request #1 from mind-gravity/localrepo

第一课作业

brightq reviewed (changes requested):
作业中需要增加updateEmployeeAddress 和UpdateSalary的两个功能 可以参考一下origin的文件然后基于那个做改动。

mind-gravity commented:
为了验证自己的学习成果，特意不看参考答案。

从解决问题的角度来说，参考答案是用了2步实现可变地址和可变薪水。第一步用updateEmployee函数获取参数address e, uint s，再将参数已赋值的变量给getPaid函数进行支付。那么我秉着节约gas的精神，直接让getpaid函数去获取参数不就好了么，何必分2步走呢？从客户端体验角度来说，都是填入2个参数，点一下鼠标，完成领薪水。流程上是一模一样，结果上也是一模一样。唯一的不同，我的code更节约gas。

另外从产品逻辑角度来说，课件中老董老师实现的功能是只有调用函数的msg.sender（frank）才有权给自己的地址领薪水，其他地址不行。那我也是按照这个思路执行的。如果说要求创建合约的人是发薪水的人，那么这件事只有用构造函数在创建合约的时候给锁定。较真点来说，第一课里面并没有介绍构造函数，这个功能属于超纲。

希望能理解。

mind-gravity committed:
根据老师的意思加入了产品逻辑的设计。

但我的想法是创建合约的人好比企业主或者是HR，在每个月核算每个员工的应发工资；
员工如果想要领工资，则领取已经核算好的部分工资。这样处理的好处是transfer权限只在员工手里。

origin文件中的核算方法是owner计算并直接发放应发计时工资，而员工则自主领取整月工资。感觉逻辑上挺混乱的不统一。
所以请不要再苛求我的代码与标准答案之前的差异了好吗？谢谢理解！

mind-gravity committed:
Merge pull request #2 from mind-gravity/localrepo

根据老师的意思加入了产品逻辑的设计

mind-gravity committed:
修复了功能异常
增加了应发工资检查器

mind-gravity committed:
Merge pull request #3 from mind-gravity/localrepo

修复了功能异常

mind-gravity committed:
完全按照官方产品意图重新实现功能。
assert和require是第二课的内容我先不用。
但是结构函数不用的话就无法实现要求，所以硬着头皮照抄，否则你叫我怎么办。

mind-gravity committed:
Merge pull request #4 from mind-gravity/localrepo

完全按照官方产品意图重新实现功能

brightq commented on Lesson-1/assignment/yours.sol:
```diff
@@ -1,39 +1,71 @@
+    
+    function updateEmployee(address e, uint s) {
+        //只有合约所有者才有权更新地址和薪水，否则就跳出
+        if (msg.sender != owner) { 
```
除了comments，似乎只是改动了 assert 和 require 变成了用If

mind-gravity replied:
想要看原创的东西可以看之前几次被reject的commit。尤其是前一次commit，跟答案完全不一样的产品逻辑。我想自己用所学实现一些新功能新业务逻辑却被告知跟参考答案不符，我照抄答案又被吐槽……所以我有点迷茫到底是要考察学生什么。我可能还是太初级了。

brightq reviewed (approved):
其实可以把updateEmployee分开来写，或者完善下其它的功能，比如如果pay之前check以下balance之类。Approved
